estimagic/optimization/process_arguments.py

```diff
@@ -0,0 +1,311 @@
+import copy
+from collections import Callable
+
+import pandas as pd
+
+
+def process_optimization_arguments(
+    criterion,
+    params,
+    algorithm,
+    criterion_kwargs=None,
+    constraints=None,
+    general_options=None,
+    algo_options=None,
+    dashboard=False,
+    db_options=None,
+):
+    """Process and validate arguments for minimize or maximize.
+
+    Args:
+        criterion (function or list of functions):
+            Python function that takes a pandas DataFrame with parameters as the first
+            argument and returns a scalar floating point value.
+
+        params (pd.DataFrame or list of pd.DataFrames):
+            See :ref:`params`.
+
+        algorithm (str or list of strings):
+            specifies the optimization algorithm. See :ref:`list_of_algorithms`.
+
+        criterion_kwargs (dict or list of dicts):
+            additional keyword arguments for criterion
+
+        constraints (list or list of lists):
+            list with constraint dictionaries. See for details.
+
+        general_options (dict):
+            additional configurations for the optimization
+
+        algo_options (dict or list of dicts):
+            algorithm specific configurations for the optimization
+
+        dashboard (bool):
+            whether to create and show a dashboard
+
+        db_options (dict):
+            dictionary with kwargs to be supplied to the run_server function.
```

Comment:
This docstring needs a `Returns:` section. I did not write it myself because it depends on your answer to my suggestion of moving the transpose step into `process_arguments`.